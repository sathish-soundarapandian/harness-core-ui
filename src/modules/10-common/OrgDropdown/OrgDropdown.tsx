/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { DropDown, SelectOption } from '@harness/uicore'
import React, { useState } from 'react'
import { useParams } from 'react-router-dom'
import { getOrganizationListPromise, OrganizationResponse } from 'services/cd-ng'
import { useStrings } from 'framework/strings'
import type { AccountPathProps } from '@common/interfaces/RouteInterfaces'

interface OrgDropdownProps {
  onChange: (item: SelectOption) => void
  value?: SelectOption
  className?: string
}

const OrgDropdown: React.FC<OrgDropdownProps> = props => {
  const { accountId } = useParams<AccountPathProps>()
  const [query, setQuery] = useState<string>()
  const { getString } = useStrings()

  function orgListPromise(): Promise<SelectOption[]> {
    return new Promise<SelectOption[]>(resolve => {
      getOrganizationListPromise({ queryParams: { accountIdentifier: accountId, searchTerm: query, pageSize: 100 } })
        .then(result => {
          let selectItems: Array<SelectOption> = []

          if (result?.data?.content?.length) {
            selectItems = result?.data?.content?.reduce?.(
              (selected: Array<SelectOption>, item: OrganizationResponse) => {
                if (item.organization?.name && item.organization?.identifier) {
                  return [...selected, { label: item.organization.name, value: item.organization.identifier }]
                }
                return selected
              },
              []
            )
          }

          resolve(selectItems)
        })
        .catch(() => {
          resolve([])
        })
    })
  }

  return (
    <DropDown
      className={props.className}
      buttonTestId="org-select"
      onChange={props.onChange}
      value={props.value}
      items={orgListPromise}
      usePortal={true}
      addClearBtn={true}
      query={query}
      onQueryChange={setQuery}
      placeholder={getString('orgsText')}
    />
  )
}

export default OrgDropdown
