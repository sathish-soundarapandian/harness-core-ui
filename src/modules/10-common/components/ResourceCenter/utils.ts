/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import moment from 'moment'
import type { UserInfo } from 'services/cd-ng'

export const ON_PREM_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/556wy85kbo-harness-on-prem-release-notes'
export const SAAS_RELEASE_NODE_LINK = 'https://ngdocs.harness.io/article/7zkchy5lhj-harness-saa-s-release-notes-2022'

export const HARNESS_SEARCH_LINK = 'https://harness.io/search/'
export const HARNESS_UNIVERISITY_LINK = 'https://university.harness.io/page/job-role-certifications'
export const HARNESS_COMMUNITY_LINK = 'https://community.harness.io/'
export const HARNESS_COMMUNITY_SLACK_LINK =
  'https://join.slack.com/t/harnesscommunity/shared_invite/zt-y4hdqh7p-RVuEQyIl5Hcx4Ck8VCvzBw'
export const HARNESS_DOCS_LINK = 'https://docs.harness.io/'
export const HARNESS_API_DOCS_LINK = 'https://apidocs.harness.io/'
export const HARNESS_TUTORIALS = 'https://developer.harness.io/tutorials/get-started'
export const SITE_STATUS_LINK = 'https://status.harness.io/'

export const timestamp = moment.now()
export const HARNESS_SUPPORT_LINK =
  '/sso.html?action=login&brand_id=114095000394&locale_id=1&return_to=https%3A%2F%2Fsupport.harness.io%2Fhc%2Fen-us%2Frequests&src=zendesk&timestamp=' +
  timestamp

export const WHATS_NEW_LINK = `https://docs.harness.io/article/ueeiji09do-what-s-new`
export const EARLY_ACCESS_LINK = `https://docs.harness.io/article/w4krvu96i3-early-access`

export const openZendeskSupport = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(HARNESS_SUPPORT_LINK)
}

export function getReleaseNodeLink(): string {
  switch (window.deploymentType) {
    case 'COMMUNITY':
    case 'ON_PREM': {
      return ON_PREM_RELEASE_NODE_LINK
    }
    default:
      return SAAS_RELEASE_NODE_LINK
  }
}
export const openFileATicket = (
  e: React.MouseEvent<Element, MouseEvent>,
  currentUserInfo: UserInfo,
  setShow?: React.Dispatch<React.SetStateAction<boolean>>
): void => {
  e.stopPropagation()
  e.preventDefault()
  window.Saber.do?.('set_options', {
    feedback_values: {
      Email: currentUserInfo.email // set default value for email field
    }
  })
  setShow?.(false)
  window.Saber.do?.('open')
}
export const openWhatsNew = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(WHATS_NEW_LINK)
}
export const openEarlyAccess = (e: React.MouseEvent<Element, MouseEvent>): void => {
  e.stopPropagation()
  e.preventDefault()
  window.open(EARLY_ACCESS_LINK)
}
