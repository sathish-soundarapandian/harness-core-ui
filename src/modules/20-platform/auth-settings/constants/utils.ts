/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { windowLocationUrlPartBeforeHash } from 'framework/utils/WindowLocation'

export const getSamlEndpoint = (accountId: string): string => {
  let url = `${windowLocationUrlPartBeforeHash()?.replace('/ng/', '')}`

  if (window.apiUrl) {
    if (window.apiUrl.startsWith('/')) {
      url = `${url}${window.apiUrl}`
    } else {
      url = window.apiUrl
    }
  }

  return `${url}/api/users/saml-login?accountId=${accountId}`
}
