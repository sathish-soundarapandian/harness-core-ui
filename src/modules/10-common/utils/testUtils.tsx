/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React from 'react'
import { UseGetProps, UseGetReturn, RestfulProvider } from 'restful-react'
import { act, fireEvent, getByText, queryByAttribute, waitFor } from '@testing-library/react'
import { compile } from 'path-to-regexp'
import { createMemoryHistory } from 'history'
import { Router, Route, Switch, useLocation, useHistory } from 'react-router-dom'
import { ModalProvider } from '@harness/use-modal'
import qs from 'qs'
import { noop } from 'lodash-es'
import { enableMapSet } from 'immer'
import userEvent from '@testing-library/user-event'
import { AppStoreContext, AppStoreContextProps } from 'framework/AppStore/AppStoreContext'
import { LicenseStoreContext, LicenseStoreContextProps } from 'framework/LicenseStore/LicenseStoreContext'
import { LICENSE_STATE_VALUES } from 'framework/LicenseStore/licenseStoreUtil'
import { withAccountId, accountPathProps } from '@common/utils/routeUtils'
import type { Project } from 'services/cd-ng'
import { StringsContext, UseStringsReturn } from 'framework/strings'

import { FeaturesContext, FeaturesContextProps } from 'framework/featureStore/FeaturesContext'
import type { FeatureDetail, FeatureMetaData } from 'framework/featureStore/featureStoreUtil'
import type { FeatureIdentifier } from 'framework/featureStore/FeatureIdentifier'
import './testUtils.scss'
import { PermissionsContext, PermissionsContextProps } from 'framework/rbac/PermissionsContext'
import { Editions } from '@common/constants/SubscriptionTypes'
import type { FeatureFlag } from '@common/featureFlags'
import { PreferenceStoreContext } from 'framework/PreferenceStore/PreferenceStoreContext'
import {
  TemplateSelectorContext,
  TemplatesSelectorContextInterface
} from 'framework/Templates/TemplateSelectorContext/TemplateSelectorContext'

export type UseGetMockData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Required<
  UseGetProps<TData, TError, TQueryParams, TPathParams>
>['mock']

export interface UseGetMockDataWithMutateAndRefetch<T> extends UseGetMockData<T> {
  mutate: () => Record<string, unknown>
  refetch: () => Record<string, unknown>
  cancel?: () => Record<string, unknown>
}

export interface UseMutateMockData<TData, TRequestBody = unknown> {
  loading?: boolean
  mutate?: (data?: TRequestBody) => Promise<TData>
}

export type UseGetReturnData<TData, TError = undefined, TQueryParams = undefined, TPathParams = undefined> = Omit<
  UseGetReturn<TData, TError, TQueryParams, TPathParams>,
  'absolutePath' | 'cancel' | 'response'
>

export const findDialogContainer = (): HTMLElement | null => document.querySelector('.bp3-dialog')
export const findPopoverContainer = (): HTMLElement | null => document.querySelector('.bp3-popover-content')

export interface TestWrapperProps {
  path?: string
  pathParams?: Record<string, string | number>
  queryParams?: Record<string, unknown>
  defaultAppStoreValues?: Partial<AppStoreContextProps>
  defaultLicenseStoreValues?: Partial<LicenseStoreContextProps>
  defaultPermissionValues?: Partial<PermissionsContextProps>
  defaultTemplateSelectorValues?: Partial<TemplatesSelectorContextInterface>
  defaultFeaturesValues?: Partial<FeaturesContextProps>
  defaultFeatureFlagValues?: Partial<Record<FeatureFlag, boolean>>
  projects?: Project[]
  enableBrowserView?: boolean
  stringsData?: Record<string, string>
  getString?: UseStringsReturn['getString']
}

export const prependAccountPath = (path: string): string => withAccountId(() => path)(accountPathProps)

export const CurrentLocation = (): JSX.Element => {
  const location = useLocation()
  return (
    <div>
      <h1>Not Found</h1>
      <div data-testid="location">{`${location.pathname}${
        location.search ? `?${location.search.replace(/^\?/g, '')}` : ''
      }`}</div>
    </div>
  )
}

export interface BrowserViewProps {
  enable?: boolean
  children: React.ReactNode
}

export function BrowserView(props: BrowserViewProps): React.ReactElement {
  const { enable, children } = props
  const location = useLocation()
  const history = useHistory()

  if (!enable) {
    return <>{children}</>
  }

  function handlePathChange(e: React.ChangeEvent<HTMLInputElement>): void {
    history.replace(e.currentTarget.value)
  }

  const search = location.search ? `?${location.search.replace(/^\?/, '')}` : ''

  return (
    <div className="browser">
      <div className="browser-header">
        <input className="browser-path" value={location.pathname + search} onChange={handlePathChange} />
      </div>
      <div className="browser-content">{children}</div>
    </div>
  )
}

export const TestWrapper: React.FC<TestWrapperProps> = props => {
  enableMapSet()
  const {
    path = '/',
    pathParams = {},
    defaultAppStoreValues,
    queryParams = {},
    defaultLicenseStoreValues,
    defaultPermissionValues,
    defaultTemplateSelectorValues,
    defaultFeaturesValues,
    defaultFeatureFlagValues = {},
    stringsData = {},
    getString = (key: string) => key
  } = props

  const search = qs.stringify(queryParams, { addQueryPrefix: true })
  const routePath = compile(path)(pathParams) + search

  // eslint-disable-next-line react-hooks/exhaustive-deps
  const history = React.useMemo(() => createMemoryHistory({ initialEntries: [routePath] }), [])

  const defaultReturn = {
    enabled: true
  }

  /** TODO: Try fixing this later. This is causing some tests to fail */
  // React.useEffect(() => {
  //   history.replace(compile(path)(pathParams) + qs.stringify(queryParams, { addQueryPrefix: true }))
  //   // eslint-disable-next-line react-hooks/exhaustive-deps
  // }, [path, pathParams, queryParams])

  return (
    <Router history={history}>
      <StringsContext.Provider value={{ data: stringsData as any, getString }}>
        <PreferenceStoreContext.Provider
          value={{
            set: noop,
            get: noop,
            clear: noop
          }}
        >
          <AppStoreContext.Provider
            value={{
              featureFlags: {
                FEATURE_ENFORCEMENT_ENABLED: true,
                ...defaultFeatureFlagValues
              },
              updateAppStore: () => void 0,
              currentUserInfo: { uuid: '' },
              ...defaultAppStoreValues
            }}
          >
            <LicenseStoreContext.Provider
              value={{
                versionMap: {},
                licenseInformation: {},
                CI_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                FF_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CCM_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CD_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CHAOS_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                STO_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CV_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                CET_LICENSE_STATE: LICENSE_STATE_VALUES.ACTIVE,
                updateLicenseStore: () => void 0,
                ...defaultLicenseStoreValues
              }}
            >
              <PermissionsContext.Provider
                value={{
                  permissions: new Map<string, boolean>(),
                  requestPermission: () => void 0,
                  checkPermission: () => true,
                  cancelRequest: () => void 0,
                  ...defaultPermissionValues
                }}
              >
                <FeaturesContext.Provider
                  value={{
                    features: new Map<FeatureIdentifier, FeatureDetail>(),
                    featureMap: new Map<FeatureIdentifier, FeatureMetaData>(),
                    getEdition: () => {
                      return undefined
                    },
                    requestFeatures: () => void 0,
                    requestLimitFeature: () => void 0,
                    checkFeature: () => {
                      return defaultReturn
                    },
                    checkLimitFeature: () => {
                      return defaultReturn
                    },
                    getRestrictionType: () => {
                      return undefined
                    },
                    getHighestEdition: () => {
                      return Editions.FREE
                    },
                    ...defaultFeaturesValues
                  }}
                >
                  <ModalProvider>
                    <TemplateSelectorContext.Provider
                      value={{
                        state: {
                          isDrawerOpened: false,
                          selectorData: {
                            templateType: 'Step',
                            onSubmit: jest.fn(),
                            onCancel: jest.fn()
                          }
                        },
                        openTemplateSelector: jest.fn(),
                        closeTemplateSelector: jest.fn(),
                        ...defaultTemplateSelectorValues
                      }}
                    >
                      <RestfulProvider base="/">
                        <BrowserView enable={props.enableBrowserView}>
                          <Switch>
                            <Route exact path={path}>
                              {props.children}
                            </Route>
                            <Route>
                              <CurrentLocation />
                            </Route>
                          </Switch>
                        </BrowserView>
                      </RestfulProvider>
                    </TemplateSelectorContext.Provider>
                  </ModalProvider>
                </FeaturesContext.Provider>
              </PermissionsContext.Provider>
            </LicenseStoreContext.Provider>
          </AppStoreContext.Provider>
        </PreferenceStoreContext.Provider>
      </StringsContext.Provider>
    </Router>
  )
}

export const queryByNameAttribute = (name: string, container: HTMLElement): HTMLElement | null =>
  queryByAttribute('name', container, name)

export const doConfigureOptionsTesting = async (
  cogModal: HTMLElement,
  fieldElement: HTMLInputElement
): Promise<void> => {
  // Type regex and submit
  // check if field has desired value
  await waitFor(() => expect(getByText(cogModal, 'common.configureOptions.regex')).toBeInTheDocument())
  const regexRadio = getByText(cogModal, 'common.configureOptions.regex')
  userEvent.click(regexRadio)
  const regexTextArea = queryByAttribute('name', cogModal, 'regExValues') as HTMLInputElement
  act(() => {
    fireEvent.change(regexTextArea, { target: { value: '<+input>.includes(/test/)' } })
  })
  const cogSubmit = getByText(cogModal, 'submit')
  userEvent.click(cogSubmit)
  await waitFor(() => expect(fieldElement.value).toBe('<+input>.regex(<+input>.includes(/test/))'))
}
