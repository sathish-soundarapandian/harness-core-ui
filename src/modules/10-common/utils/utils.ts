/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { getMultiTypeFromValue, IconName, MultiSelectOption, MultiTypeInputType, SelectOption } from '@harness/uicore'
import { defaultTo, isUndefined } from 'lodash-es'
import { Scope } from '@common/interfaces/SecretsInterface'
import { ModuleName, moduleToModuleNameMapping } from 'framework/types/ModuleName'
import { useLicenseStore } from 'framework/LicenseStore/LicenseStoreContext'
import { Editions } from '@common/constants/SubscriptionTypes'
import type { UserMetadataDTO } from 'services/cd-ng'
import type { Module } from '@common/interfaces/RouteInterfaces'
import { useModuleInfo } from '@common/hooks/useModuleInfo'

const PR_ENV_HOST_NAME = 'pr.harness.io'

interface SetPageNumberProps {
  setPage: (value: number) => void
  pageItemsCount?: number
  page: number
}

export const getModuleIcon = (module: ModuleName): IconName => {
  switch (module) {
    case ModuleName.CD:
      return 'cd-main'
    case ModuleName.CV:
      return 'cv-main'
    case ModuleName.CI:
      return 'ci-main'
    case ModuleName.CE:
      return 'ce-main'
    case ModuleName.CF:
      return 'cf-main'
    case ModuleName.STO:
      return 'sto-color-filled'
    case ModuleName.CHAOS:
      return 'chaos-main'
    case ModuleName.ET:
      return 'cet'
  }
  return 'nav-project'
}

export const getReference = (scope?: Scope, identifier?: string): string | undefined => {
  switch (scope) {
    case Scope.PROJECT:
      return identifier
    case Scope.ORG:
      return `org.${identifier}`
    case Scope.ACCOUNT:
      return `account.${identifier}`
  }
}

export const setPageNumber = ({ setPage, pageItemsCount, page }: SetPageNumberProps): void => {
  if (pageItemsCount === 0 && page > 0) {
    setPage(page - 1)
  }
}

export const formatCount = (num: number): string => {
  const min = 1e3
  if (num >= min) {
    const units = ['k', 'M', 'B', 'T']
    const order = Math.floor(Math.log(num) / Math.log(1000))
    const finalNum = Math.round(num / 1000 ** order)
    if (finalNum === min && order < units.length) {
      return 1 + units[order]
    }
    return finalNum + units[order - 1]
  }
  return num.toLocaleString()
}

export function useGetCommunity(): boolean {
  const { licenseInformation } = useLicenseStore()
  return licenseInformation?.['CD']?.edition === Editions.COMMUNITY
}

export const isOnPrem = (): boolean => window.deploymentType === 'ON_PREM'

export const focusOnNode = (node: HTMLElement): void => {
  const oldTabIndex = node.tabIndex
  node.tabIndex = -1
  node.focus()
  node.tabIndex = oldTabIndex
}

const HOTJAR_SUPPRESSION_ATTR = 'data-hj-suppress'

// Utility to add `data-hj-suppress` into a collection of elements to
// suppress data from HotJar recording
// @see https://bit.ly/3rCgpOY
export const suppressHotJarRecording = (elements: Element[] | null | undefined): void => {
  if (window.hj) {
    elements?.forEach?.((e: Element) => {
      if (!e.hasAttribute(HOTJAR_SUPPRESSION_ATTR)) {
        e.setAttribute(HOTJAR_SUPPRESSION_ATTR, 'true')
      }
    })
  }
}

// Utility to generate { 'data-hj-suppress': true } attribute if HotJar is available
export const addHotJarSuppressionAttribute = (): { [HOTJAR_SUPPRESSION_ATTR]: boolean } | undefined =>
  window.hj ? { [HOTJAR_SUPPRESSION_ATTR]: true } : undefined

// Utility to check if environment is a PR environment
export const isPR = (): boolean => {
  return location.hostname?.includes(PR_ENV_HOST_NAME)
}

// Utility to check if environment is a local develop environment
export const isLocalHost = (): boolean => {
  return location.hostname === 'localhost' || location.hostname === '127.0.0.1'
}

export const getPREnvNameFromURL = (url: string): string => {
  if (!url) {
    return ''
  }
  return isPR() ? url.split(PR_ENV_HOST_NAME)?.[1]?.split('/')?.[1] : ''
}

export function isMultiTypeFixed(type: MultiTypeInputType): boolean {
  return type === MultiTypeInputType.FIXED
}

export function isValueFixed(
  value?: boolean | string | number | SelectOption | string[] | MultiSelectOption[]
): boolean {
  const type = getMultiTypeFromValue(value)

  return isMultiTypeFixed(type)
}

export function isMultiTypeRuntime(type: MultiTypeInputType): boolean {
  return [MultiTypeInputType.EXECUTION_TIME, MultiTypeInputType.RUNTIME].includes(type)
}

export function isValueRuntimeInput(
  value?: boolean | string | number | SelectOption | string[] | MultiSelectOption[]
): boolean {
  const type = getMultiTypeFromValue(value)

  return isMultiTypeRuntime(type)
}

export function isMultiTypeExpression(type: MultiTypeInputType): boolean {
  return type === MultiTypeInputType.EXPRESSION
}

export function getLastURLPathParam(URL: string): string {
  const splitArray = URL.split('/')
  return splitArray.length > 0 ? splitArray[splitArray.length - 1]?.split('-')?.join('') : ''
}

export function isValueExpression(
  value?: boolean | string | number | SelectOption | string[] | MultiSelectOption[]
): boolean {
  const type = getMultiTypeFromValue(value)

  return isMultiTypeExpression(type)
}

export const getUserName = (user: UserMetadataDTO): string => {
  return defaultTo(user.name, user.email)
}

export const REFERER_URL = 'refererURL'

export const getSavedRefererURL = (): string => localStorage.getItem(REFERER_URL) || ''
export const getGaClientID = (): string => {
  try {
    return defaultTo(
      document?.cookie
        ?.split('; ')
        ?.find((key: string) => key.includes('_ga='))
        ?.split('.')
        ?.slice(2)
        ?.join('.'),
      ''
    )
  } catch (e) {
    return ''
  }
}

export const isWindowsOS = (): boolean => navigator?.appVersion?.indexOf('Win') != -1

export const toBase64 = (file: File): Promise<string> =>
  new Promise((resolve, reject) => {
    const reader = new FileReader()
    reader.onloadend = () => resolve((reader.result || '').toString())
    reader.onerror = reject
    reader.readAsDataURL(file)
  })

export const getImageDimensions = (url: string): Promise<{ width: number; height: number }> => {
  return new Promise((resolve, reject) => {
    const img = document.createElement('img')

    img.onload = () => {
      const width = img.naturalWidth
      const height = img.naturalHeight

      resolve({ width, height })
    }
    img.onerror = reject
    img.src = url
  })
}

export interface NumberFormatterOptions {
  truncate?: boolean
}

export const numberFormatter: (value?: number, options?: NumberFormatterOptions) => string = (
  value?: number,
  options = { truncate: true }
) => {
  // istanbul ignore if
  if (value === undefined) {
    return ''
  }
  const truncateOptions = [
    { value: 1000000, suffix: 'm' },
    { value: 1000, suffix: 'k' }
  ]
  if (options.truncate) {
    for (const truncateOption of truncateOptions) {
      if (value >= truncateOption.value) {
        const truncatedValue = value / truncateOption.value
        // istanbul ignore if
        if (truncatedValue % 1 !== 0) {
          return `${truncatedValue.toFixed(1)}${truncateOption.suffix}`
        }
        return `${truncatedValue}${truncateOption.suffix}`
      }
    }
  }
  return `${getFixed(value)}`
}

export const getFixed = (value: number, places = 1): number => {
  if (value % 1 === 0) {
    return value
  }
  return parseFloat(value.toFixed(places))
}

export const getRefFromIdentifier = (
  identifier: string,
  orgIdentifier?: string,
  projectIdentifier?: string
): string => {
  if (projectIdentifier) {
    return identifier
  } else if (!orgIdentifier) {
    return `account.${identifier}`
  }
  return `org.${identifier}`
}

export const getIdentifierFromScopedRef = (entity: string): string => {
  return entity.indexOf('.') < 0 ? entity : entity.split('.')[1]
}

export const countAllKeysInObject = (obj: Record<string, any>): number =>
  Object.keys(obj).reduce((acc, curr) => {
    if (typeof obj[curr] === 'object' && !Array.isArray(obj[curr]) && obj[curr] !== null) {
      return 1 + acc + countAllKeysInObject(obj[curr])
    } else {
      return 1 + acc
    }
  }, 0)

export const isSimplifiedYAMLEnabled = (module?: Module, isSimpliedYAMLFFEnabled?: boolean): boolean => {
  return isSimpliedYAMLFFEnabled
    ? module?.valueOf().toLowerCase() === moduleToModuleNameMapping.ci.toLowerCase() ||
        module?.valueOf().toLowerCase() === moduleToModuleNameMapping.iacm.toLowerCase()
    : false
}

export function useGetCDFree(): boolean {
  const { licenseInformation } = useLicenseStore()
  return licenseInformation?.['CD']?.edition === Editions.FREE
}

export const useGetFreeOrCommunityCD = (): boolean => {
  const { module } = useModuleInfo()
  const isCommunity = useGetCommunity()
  const isFree = useGetCDFree()
  // Visible at projects and cd all empty state pages
  return (module === 'cd' || isUndefined(module)) && (isFree || isCommunity)
}
