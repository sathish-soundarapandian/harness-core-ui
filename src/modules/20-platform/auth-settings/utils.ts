/*
 * Copyright 2022 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { get } from 'lodash-es'
import type { CreateIpAllowlistConfigProps, UpdateIpAllowlistConfigProps } from '@harnessio/react-ng-manager-client'

import type { IIPAllowlistForm } from '@auth-settings/interfaces/IPAllowlistInterface'
import { SourceType } from '@auth-settings/interfaces/IPAllowlistInterface'
import { CurrencyType } from '@common/constants/SubscriptionTypes'

export function getAmountInCurrency(currency: CurrencyType, amount: number): string {
  switch (currency) {
    case CurrencyType.USD: {
      return amount.toLocaleString('en-US', {
        style: 'currency',
        currency: 'USD'
      })
    }
  }

  return ''
}

// dividing the actual amount by 100 as amount numbers are in cents
export function getDollarAmount(amount?: number, isYearly?: boolean): number {
  if (isYearly) {
    return (amount || 0) / 1200
  }
  return (amount || 0) / 100
}

export const getErrorMessage = (error: any): string =>
  get(error, 'data.error', get(error, 'data.message', error?.message))

export const buildCreateIPAllowlistPayload = (data: IIPAllowlistForm): CreateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: false, // always false while creating
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags
      }
    }
  }
}

export const buildUpdateIPAllowlistPayload = (
  data: IIPAllowlistForm,
  overrideFields?: any
): UpdateIpAllowlistConfigProps => {
  const allowSourceType = []
  if (data.allowSourceTypeUI) {
    allowSourceType.push(SourceType.UI)
  }
  if (data.allowSourceTypeAPI) {
    allowSourceType.push(SourceType.API)
  }
  return {
    'ip-config-identifier': data.identifier,
    body: {
      ip_allowlist_config: {
        allowed_source_type: allowSourceType,
        description: data?.description,
        enabled: data?.enabled,
        identifier: data.identifier,
        ip_address: data.ipAddress,
        name: data.name,
        tags: data?.tags,
        ...overrideFields
      }
    }
  }
}
