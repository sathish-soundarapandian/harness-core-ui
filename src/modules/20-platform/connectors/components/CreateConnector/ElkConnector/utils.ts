/*
 * Copyright 2021 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import { ElkAuthType } from '@connectors/constants'
import type { FormData } from '@connectors/interfaces/ConnectorInterface'
import type { ConnectorConfigDTO } from 'services/cd-ng'

type InitializeElkConnectorArgs = {
  prevStepData?: ConnectorConfigDTO
  accountId: string
  projectIdentifier?: string
  orgIdentifier?: string
}

export function initializeElkConnector({
  prevStepData,
  accountId,
  projectIdentifier,
  orgIdentifier
}: InitializeElkConnectorArgs): FormData {
  const defaultObj = {
    url: '',
    accountId,
    username: '',
    authType: ElkAuthType.USERNAME_PASSWORD,
    password: undefined,
    projectIdentifier,
    orgIdentifier
  }

  if (!prevStepData) {
    return defaultObj
  }
  const { spec, ...prevData } = prevStepData
  return {
    ...defaultObj,
    url: prevData?.url || spec?.controllerUrl || spec?.url || '',
    password: prevData?.password || spec?.passwordRef,
    clientSecretRef: prevData?.clientSecretRef || spec?.clientSecretRef,
    authType: prevData?.authType || spec?.authType || ElkAuthType.USERNAME_PASSWORD,
    username: prevData?.username || spec?.username || '',
    clientId: prevData?.clientId || spec?.clientId,
    apiKeyId: prevData?.apiKeyId || spec?.apiKeyId,
    apiKeyRef: prevData?.apiKeyRef || spec?.apiKeyRef
  }
}
