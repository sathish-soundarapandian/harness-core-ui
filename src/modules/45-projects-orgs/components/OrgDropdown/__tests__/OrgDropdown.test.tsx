/*
 * Copyright 2023 Harness Inc. All rights reserved.
 * Use of this source code is governed by the PolyForm Shield 1.0.0 license
 * that can be found in the licenses directory at the root of this repository, also available at
 * https://polyformproject.org/wp-content/uploads/2020/06/PolyForm-Shield-1.0.0.txt.
 */

import React, { useEffect, useState } from 'react'
import { render } from '@testing-library/react'
import type { SelectOption } from '@harness/uicore'
import { noop } from 'lodash-es'
import { TestWrapper } from '@common/utils/testUtils'
import { OrgMockData } from '@projects-orgs/pages/projects/__tests__/ProjectPageMock'
import OrgDropdown from '../OrgDropdown'

jest.mock('@harness/uicore', () => {
  return {
    ...jest.requireActual('@harness/uicore'),
    DropDown: (props: any) => {
      const [items, setItems] = useState<SelectOption[]>([])

      useEffect(() => {
        props.items().then((itemsResponse: SelectOption[]) => {
          setItems(itemsResponse)
        })
      }, [])

      return (
        <div>
          orgDropdown
          <button data-testid="onChangeBtn" />
          <div data-testid="dropdownItems">
            {items.map((item: SelectOption) => (
              <div key={item.value as string}>{item.value}</div>
            ))}
          </div>
        </div>
      )
    }
  }
})

let orgListPromiseMock = jest.fn().mockImplementation(() => {
  return Promise.resolve(OrgMockData)
})

jest.mock('services/cd-ng', () => ({
  getOrganizationListPromise: jest.fn().mockImplementation(() => {
    return orgListPromiseMock()
  })
}))

describe('org dropdown test', () => {
  test('render', () => {
    const { container } = render(
      <TestWrapper>
        <OrgDropdown onChange={noop} />
      </TestWrapper>
    )

    expect(container).toMatchSnapshot()
  })

  test('test when data is not defined', () => {
    orgListPromiseMock = jest.fn().mockImplementation(() => {
      return Promise.resolve(undefined)
    })
    const { queryByText } = render(
      <TestWrapper>
        <OrgDropdown onChange={noop} />
      </TestWrapper>
    )
    expect(queryByText('orgDropdown')).not.toBeNull()
  })
})
