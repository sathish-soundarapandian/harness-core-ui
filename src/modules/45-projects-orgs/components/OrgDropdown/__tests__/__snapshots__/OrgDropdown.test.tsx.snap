// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`org dropdown test render 1`] = `
<div>
  <div>
    orgDropdown
    <button
      data-testid="onChangeBtn"
    />
    <div
      data-testid="dropdownItems"
    />
  </div>
</div>
`;
